--- /home/maier/navit_linux/svn-5214/src/navit/roadprofile.c	2012-09-04 20:04:37.387424011 +0200
+++ roadprofile.c	2012-08-23 23:11:00.199795000 +0200
@@ -35,6 +35,11 @@
 	case attr_route_weight:
 		this->route_weight=attr->u.num;
 		break;
+/***************************** Rainer ***************************/
+	case attr_zoom_scale:
+		this->zoom_scale=attr->u.num;
+		break;
+/***************************** Rainer ***************************/
 	default:
 		break;
 	}
@@ -48,6 +53,9 @@
 	this_=g_new0(struct roadprofile, 1);
 	this_->attrs=attr_list_dup(attrs);
 	this_->maxspeed=0;
+/***************************** Rainer ***************************/
+	this_->zoom_scale=32;
+/***************************** Rainer ***************************/
 	for (attr=attrs;*attr; attr++) 
 		roadprofile_set_attr_do(this_, *attr);
 	return this_;
