Index: navit/navit/maptool/osm.c
===================================================================
--- navit/navit/maptool/osm.c	(revision 2895)
+++ navit/navit/maptool/osm.c	(working copy)
@@ -92,6 +92,8 @@
 	{ 36,"Australia,AUS"},
 	{ 40,"Austria,Österreich,AUT"},
 	{ 56,"Belgium"},
+	{ 70,"Bosnia and Herzegovina,Bosna i Hercegovina,Босна и Херцеговина"},
+	{100,"Bulgaria,България"},
 	{124,"Canada"},
 	{152,"Chile"},
 	{191,"Croatia,Republika Hrvatska,HR"},
@@ -104,16 +106,19 @@
 	{348,"Hungary"},
  	{380,"Italy,Italia"},
 	{442,"Luxembourg"},
+	{499,"Montenegro,Црна Гора,Crna Gora"},
 	{528,"Nederland,The Netherlands,Niederlande,NL"},
 	{578,"Norway,Norge,Noreg,NO"},
 	{616,"Poland,Polska,PL"},
 	{642,"România,Romania,RO"},
 	{643,"Россия,Российская Федерация,Russia,Russian Federation"},
+	{688,"Srbija,Србија,Serbia"},
 	{703,"Slovakia,Slovensko,SK"},
 	{705,"Slovenia,Republika Slovenija,SI"},
  	{724,"Spain,Espana,España,Reino de Espana"},
 	{752,"Sweden,Sverige,Konungariket Sverige,SE"},
 	{756,"Schweiz"}, 
+	{807,"Macedonia,Македонија"},
 	{826,"United Kingdom,UK"},
 	{840,"USA"},
 	{999,"Unknown"},
