diff --git a/navit/file.c b/navit/file.c
index bee93b7..f02c017 100644
--- a/navit/file.c
+++ b/navit/file.c
@@ -145,7 +145,7 @@ int file_mkdir(char *name, int pflag)
 int
 file_mmap(struct file *file)
 {
-	int mmap_size=file->size+1024*1024;
+	int mmap_size=file->size;
 #ifdef HAVE_API_WIN32_BASE
 	file->begin = (char*)mmap_readonly_win32( file->name, &file->map_handle, &file->map_file );
 #else
