diff --git a/navit/navit/graphics.c b/navit/navit/graphics.c
index 5e722f8..54b3a86 100644
--- a/navit/navit/graphics.c
+++ b/navit/navit/graphics.c
@@ -1000,8 +1000,12 @@ static int
 int_sqrt(unsigned int n)
 {
 	unsigned int h, p= 0, q= 1, r= n;
-	while ( q <= n )
+	while ( q <= n ) {
 		q <<= 2;
+		if(q == 0) {
+			return (int) sqrtf( (float) n ); /* use float sqrt if we reach q MAX */
+		}
+	}
 	while ( q != 1 ) {
 		q >>= 2;
 		h = p + q;
