From 12503c76f0768fb0a63f71fd97b7d2d88f6b5fd8 Mon Sep 17 00:00:00 2001
From: mvglasow <michael -at- vonglasow.com>
Date: Tue, 30 Sep 2014 23:27:18 +0200
Subject: [PATCH] Fix:core:Handle percentage sizes for OSD elements. Fixes
 #983 and #1098.

Signed-off-by: mvglasow <michael -at- vonglasow.com>
---
 navit/navit/osd.c |   43 ++++++++++++++++++++++++++++++++++++-------
 1 file changed, 36 insertions(+), 7 deletions(-)

diff --git a/navit/navit/osd.c b/navit/navit/osd.c
index bec0b41..00e2411 100644
--- a/navit/navit/osd.c
+++ b/navit/navit/osd.c
@@ -158,26 +158,54 @@ osd_std_resize(struct osd_item *item)
  	graphics_overlay_resize(item->gr, &item->p, item->w, item->h, 65535, 1);
 }
  
+/**
+ * @brief Calculates the size and position of an OSD item.
+ *
+ * If the geometry of the OSD item is specified relative to screen dimensions,
+ * this function will set its absolute dimensions accordingly.
+ * @param item
+ * @param w Available screen width in pixels (the width that corresponds to
+ * 100%)
+ * @param h Available screen height in pixels (the height that corresponds to
+ * 100%)
+ */
 static void
-osd_std_calculate_sizes(struct osd_item *item, struct osd_priv *priv, int w, int h) 
+osd_std_calculate_sizes(struct osd_item *item, int w, int h)
 {
-	struct attr vehicle_attr;
-
  	if (item->rel_w) {
 		item->w = (item->rel_w * w) / 100;
  	}
- 
+
  	if (item->rel_h) {
 		item->h = (item->rel_h * h) / 100;
  	}
- 
+
  	if (item->rel_x) {
 		item->p.x = (item->rel_x * w) / 100;
  	}
- 
+
  	if (item->rel_y) {
 		item->p.y = (item->rel_y * h) / 100;
  	}
+}
+
+/**
+ * @brief Recalculates the size and position of an OSD item and
+ * triggers a redraw of the item.
+ *
+ * @param item
+ * @param priv
+ * @param w Available screen width in pixels (the width that corresponds to
+ * 100%)
+ * @param h Available screen height in pixels (the height that corresponds to
+ * 100%)
+ */
+static void
+osd_std_calculate_sizes_and_redraw(struct osd_item *item, struct osd_priv *priv, int w, int h)
+{
+	struct attr vehicle_attr;
+
+	osd_std_calculate_sizes(item, w, h);
 
 	osd_std_resize(item);
 	if (item->meth.draw) {
@@ -364,6 +392,7 @@ osd_set_std_graphic(struct navit *nav, struct osd_item *item, struct osd_priv *p
 	struct graphics *navit_gr;
 
 	navit_gr = navit_get_graphics(nav);
+	osd_std_calculate_sizes(item, navit_get_width(nav), navit_get_height(nav));
 	item->gr = graphics_overlay_new(navit_gr, &item->p, item->w, item->h, 65535, 1);
 
 	item->graphic_bg = graphics_gc_new(item->gr);
@@ -381,7 +410,7 @@ osd_set_std_graphic(struct navit *nav, struct osd_item *item, struct osd_priv *p
 
 	osd_set_std_config(nav, item);
 
-	item->resize_cb = callback_new_attr_2(callback_cast(osd_std_calculate_sizes), attr_resize, item, priv);
+	item->resize_cb = callback_new_attr_2(callback_cast(osd_std_calculate_sizes_and_redraw), attr_resize, item, priv);
 	graphics_add_callback(navit_gr, item->resize_cb);
 	osd_set_keypress(nav, item);
 }
-- 
1.7.9.5

