Index: binfile.c
===================================================================
--- binfile.c	(revision 1384)
+++ binfile.c	(working copy)
@@ -119,6 +119,7 @@
 
 static void cd_to_cpu(struct zip_cd *zcd) {
 	g_assert(zcd != NULL);
+	zcd->zipcensig = le32_to_cpu(zcd->zipcensig);
 	zcd->zipccrc   = le32_to_cpu(zcd->zipccrc);
 	zcd->zipcsiz   = le32_to_cpu(zcd->zipcsiz);
 	zcd->zipcunc   = le32_to_cpu(zcd->zipcunc);
