--- navit/maptool/osm_OLD.c	2010-10-24 22:40:40.835550748 +0100
+++ navit/maptool/osm.c	2010-10-24 22:40:53.383548679 +0100
@@ -1050,6 +1050,13 @@
 		strcpy(is_in_buffer, v);
 		level=5;
 	}
+	if (! strcmp(k,"is_in:country")) {
+ 		/*	sometimes there is no is_in tag, only is_in:country. 
+		   I put this here so it can be overwritten by the next if clause if there IS an is_in tag.
+		*/
+		strcpy(is_in_buffer, v);
+		level=5;
+	}
 	if (! strcmp(k,"gnis:ST_alpha")) {
 		/*	assume a gnis tag means it is part of the USA:
 			http://en.wikipedia.org/wiki/Geographic_Names_Information_System
