Index: osm2navit.c
===================================================================
--- osm2navit.c	(revision 2349)
+++ osm2navit.c	(working copy)
@@ -398,7 +398,8 @@
 	{ 40,"Austria,Österreich,AUT"},
 	{ 56,"Belgium"},
 	{124,"Canada"},
-	{203,"Czech Republic"},
+	{191,"Croatia,Republika Hrvatska,HR"},
+	{203,"Czech Republic,Česká republika,CZ"},
  	{208,"Denmark,Danmark,DK"},
  	{246,"Finland,Suomi"},
  	{250,"France,République française,FR"},
@@ -407,6 +408,8 @@
 	{528,"Nederland,The Netherlands,Niederlande,NL"},
 	{578,"Norway,Norge,Noreg,NO"},
 	{616,"Poland,Polska,PL"},
+	{703,"Slovakia,Slovensko,SK"},
+	{705,"Slovenia,Republika Slovenija,SI"},
  	{724,"Spain,Espana,Espana,Reino de Espana"},
 	{752,"Sweden,Sverige,Konungariket Sverige,SE"},
 	{756,"Schweiz"}, 
