Index: navit/intl_cmake/CMakeLists.txt
===================================================================
--- navit/intl_cmake/CMakeLists.txt	(Revision 5496)
+++ navit/intl_cmake/CMakeLists.txt	(Arbeitskopie)
@@ -1,4 +1,4 @@
 add_definitions(-DLOCALE_ALIAS_PATH=LOCALEDIR -DIN_LIBINTL=1 -DHAVE_CONFIG_H=1 -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DLIBDIR=\"${CMAKE_INSTALL_PREFIX}/${LIB_DIR}\")
-supportlib_add_library(intl_cmake bindtextdom.c dcgettext.c dcigettext.c dcngettext.c dgettext.c dngettext.c explodename.c finddomain.c gettext.c intl-compat.c l10nflist.c loadmsgcat.c localcharset.c localealias.c localename.c log.c ngettext.c osdep.c plural-exp.c plural.c printf-args.c printf.c relocatable.c textdomain.c vasnprintf.c)
+supportlib_add_library(support_gettext_intl bindtextdom.c dcgettext.c dcigettext.c dcngettext.c dgettext.c dngettext.c explodename.c finddomain.c gettext.c intl-compat.c l10nflist.c loadmsgcat.c localcharset.c localealias.c localename.c log.c ngettext.c osdep.c plural-exp.c plural.c printf-args.c printf.c relocatable.c textdomain.c vasnprintf.c)
 ADD_CUSTOM_TARGET(libintl.h DEPENDS libgnuintl.h COMMAND ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_SOURCE_DIR}/libgnuintl.h ${CMAKE_CURRENT_BINARY_DIR}/libintl.h)
-ADD_DEPENDENCIES(intl_cmake libintl.h)
+ADD_DEPENDENCIES(support_gettext_intl libintl.h)
Index: navit/CMakeLists.txt
===================================================================
--- navit/CMakeLists.txt	(Revision 5496)
+++ navit/CMakeLists.txt	(Arbeitskopie)
@@ -155,7 +155,7 @@
 
 ADD_DEPENDENCIES(${NAVIT_LIBNAME} version)
 if (USE_LIBGNUINTL AND NOT HAVE_GLIB)
-   ADD_DEPENDENCIES(support_glib intl_cmake)
+   ADD_DEPENDENCIES(support_glib support_gettext_intl)
 endif()
 
 
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(Revision 5496)
+++ CMakeLists.txt	(Arbeitskopie)
@@ -173,7 +173,7 @@
 ### Configure build
 #Required libraries, supplied with navit
 if (NOT HAVE_LIBINTL)
-   add_plugin(intl_cmake "native libintl missing" TRUE)
+   add_plugin(support/gettext_intl "native libintl missing" TRUE)
    set(USE_LIBGNUINTL TRUE)
 else()
    CHECK_LIBRARY_EXISTS(intl libintl_fprintf "" NEEDS_LIBINTL)
