Index: route.c
===================================================================
--- route.c	(revision 2376)
+++ route.c	(working copy)
@@ -722,10 +722,10 @@
 		route_info_free(this->pos);
 	this->pos=NULL;
 	this->pos=route_find_nearest_street(this->vehicleprofile, this->ms, pos);
-	this->pos->street_direction=0;
 	dbg(1,"this->pos=%p\n", this->pos);
 	if (! this->pos)
 		return;
+    this->pos->street_direction=0;
 	route_info_distances(this->pos, pos->pro);
 	route_path_update(this, 0, 1);
 }
