Index: navit/vehicle/wince/vehicle_wince.c
===================================================================
--- navit/vehicle/wince/vehicle_wince.c	(revision 2964)
+++ navit/vehicle/wince/vehicle_wince.c	(working copy)
@@ -633,9 +633,9 @@
 	case attr_position_time_iso8601:
 		if (!priv->fixyear || !priv->fixtime[0])
 			return 0;
-		sprintf(priv->fixiso8601, "%04d-%02d-%02dT%sZ",
+		sprintf(priv->fixiso8601, "%04d-%02d-%02dT%.2s:%.2s:%sZ",
 			priv->fixyear, priv->fixmonth, priv->fixday,
-			priv->fixtime);
+						priv->fixtime, (priv->fixtime+2), (priv->fixtime+4));
 		attr->u.str=priv->fixiso8601;
 		break;
 	case attr_active:
