Index: Makefile.am
===================================================================
--- Makefile.am	(revision 3490)
+++ Makefile.am	(working copy)
@@ -84,7 +84,7 @@
 android/build.xml: android/AndroidManifest.xml
 	mkdir -p android
 	if [ "$(builddir)" != "$(srcdir)" ]; then cp -rp $(srcdir)/android/* $(builddir)/android; fi
-	android update project -t 3 -p android
+	android update project -t android-3 -p android
 
 apk: all android/build.xml navit.xml
 	mkdir -p android/libs/armeabi
