Index: osd_core.c
===================================================================
--- osd_core.c	(revision 3586)
+++ osd_core.c	(working copy)
@@ -1076,15 +1076,13 @@
             if (rprof) {
                 if(rprof->maxspeed!=0)
                     routespeed=rprof->maxspeed;
-                else
-                    routespeed=rprof->speed;
             }
         }
         tracking_get_attr(tracking, attr_position_speed, &speed_attr, NULL);
         tracking_speed = *speed_attr.u.numd;
         if( -1 != tracking_speed && -1 != routespeed ) {
             snprintf(text,16,"%.0lf",routespeed);
-            if( this->speed_exceed_limit_offset+routespeed<tracking_speed ||
+            if( this->speed_exceed_limit_offset+routespeed<tracking_speed &&
                 (100.0+this->speed_exceed_limit_percent)/100.0*routespeed<tracking_speed ) {
                 if(this->announce_state==eNoWarn && this->announce_on) {
                     this->announce_state=eWarningTold; //warning told
