Index: osd_core.c
===================================================================
--- osd_core.c	(revision 2373)
+++ osd_core.c	(working copy)
@@ -146,7 +146,7 @@
 {
 	struct point p,bbox[4];
 	struct attr attr_dir, destination_attr, position_attr;
-	double dir, vdir = 0;
+	double dir, yaw = 0;
 	char *buffer;
 	struct coord c1, c2;
 	enum projection pro;
@@ -157,10 +157,8 @@
 	graphics_draw_circle(this->osd_item.gr,
 			     this->osd_item.graphic_fg_white, &p, this->osd_item.w*5/6);
 	if (v) {
-		if (vehicle_get_attr(v, attr_position_direction, &attr_dir, NULL)) {
-			vdir = *attr_dir.u.numd;
-			handle(this->osd_item.gr, this->osd_item.graphic_fg_white, &p, this->osd_item.w/3, -vdir);
-		}
+		yaw=transform_get_yaw(navit_get_trans(nav));
+		handle(this->osd_item.gr, this->osd_item.graphic_fg_white, &p, this->osd_item.w/3, -yaw);
 
 		if (navit_get_attr(nav, attr_destination, &destination_attr, NULL)
 		    && vehicle_get_attr(v, attr_position_coord_geo,&position_attr, NULL)) {
@@ -169,7 +167,7 @@
 			c2.x = destination_attr.u.pcoord->x;
 			c2.y = destination_attr.u.pcoord->y;
 			dir = atan2(c2.x - c1.x, c2.y - c1.y) * 180.0 / M_PI;
-			dir -= vdir;
+			dir -= yaw;
 			handle(this->osd_item.gr, this->green, &p, this->osd_item.w/3, dir);
 			buffer=format_distance(transform_distance(pro, &c1, &c2),"");
 			graphics_get_text_bbox(this->osd_item.gr, this->osd_item.font, buffer, 0x10000, 0, bbox, 0);
