--- a/navit/po/uk.po.in	2009-04-12 16:12:37.000000000 +0200
+++ b/navit/po/uk.po.in	2009-04-12 16:12:57.000000000 +0200
@@ -119,7 +119,7 @@
 #, c-format
 msgid "one kilometer"
 msgid_plural "%d kilometers"
-msgstr[0] "один кілометр"
+msgstr[0] "%d один кілометр"
 msgstr[1] "%d кілометри"
 msgstr[2] "%d кілометрів"
 
@@ -127,7 +127,7 @@
 #, c-format
 msgid "in one kilometer"
 msgid_plural "in %d kilometers"
-msgstr[0] "за один кілометр"
+msgstr[0] "за %d один кілометр"
 msgstr[1] "за %d кілометри"
 msgstr[2] "за %d кілометрів"
 
