--- gui_internal_old.c	2010-08-15 10:58:52.302306139 +0100
+++ gui_internal.c	2010-09-03 23:30:22.046299644 +0100
@@ -5178,9 +5178,9 @@
 	data->button_box=gui_internal_box_new(this,
 					      gravity_center|orientation_horizontal);
 	data->button_box->children=g_list_append(data->button_box->children,
-						 data->next_button);
-	data->button_box->children=g_list_append(data->button_box->children,
 						 data->prev_button);
+	data->button_box->children=g_list_append(data->button_box->children,
+						 data->next_button);
 	//data->button_box->background=this->background2;
 	data->button_box->bl=this->spacing;
 	widget->children=g_list_append(widget->children,data->button_box);
