Index: graphics.c
===================================================================
--- graphics.c	(revision 4387)
+++ graphics.c	(working copy)
@@ -1703,6 +1703,8 @@
 	struct point p;
 	char *path;
 
+int poly_count=256;
+
 	while (di) {
 	int i,count=di->count,mindist=dc->mindist;
 
@@ -1716,12 +1718,16 @@
 	if (dc->type == type_poly_water_tiled)
 		mindist=0;
 	if (dc->e->type == element_polyline)
-		count=transform(dc->trans, dc->pro, di->c, pa, count, mindist, e->u.polyline.width, width);
-	else
-		count=transform(dc->trans, dc->pro, di->c, pa, count, mindist, 0, NULL);
+			count=transform(dc->trans, dc->pro, di->c, pa, count, mindist, e->u.polyline.width, width);
+		else
+                     if (dc->e->type != element_polygon || (dc->e->type == element_polygon && poly_count))
+		        count=transform(dc->trans, dc->pro, di->c, pa, count, mindist, 0, NULL);
 	switch (e->type) {
 	case element_polygon:
-		graphics_draw_polygon_clipped(gra, gc, pa, count);
+		if (poly_count) 
+                     {--poly_count;
+                     graphics_draw_polygon_clipped(gra, gc, pa, count);
+                     }
 		break;
 	case element_polyline:
 		{
