--- navit-0.2.0~svn2974+dfsg.1.orig/navit/binding/dbus/binding_dbus.c
+++ navit-0.2.0~svn2974+dfsg.1/navit/binding/dbus/binding_dbus.c
@@ -1426,6 +1426,7 @@
 	{".gui",     "command",            "s",       "command",                                 "a{sa{sv}}",  "return", request_gui_command},
 	{".navit",  "draw",                "",        "",                                        "",   "",      request_navit_draw},
 	{".navit",  "add_message",         "s",       "message",                                 "",   "",      request_navit_add_message},
+	{".navit",  "set_center_by_string",          "s",       "(coordinates)",                           "",   "",      request_navit_set_center},
 	{".navit",  "set_center",          "s",       "(coordinates)",                           "",   "",      request_navit_set_center},
 	{".navit",  "set_center",          "(is)",    "(projection,coordinates)",                "",   "",      request_navit_set_center},
 	{".navit",  "set_center",          "(iii)",   "(projection,longitude,latitude)",         "",   "",      request_navit_set_center},
