Index: navit/gui/internal/gui_internal.c
===================================================================
--- navit/gui/internal/gui_internal.c	(revision 2062)
+++ navit/gui/internal/gui_internal.c	(working copy)
@@ -2735,10 +2735,18 @@
 {
 	if (this->sl)
 		search_list_select(this->sl, attr_country_all, 0, 0);
-	gui_internal_search(this,_("Town"),"Town",1);
+	gui_internal_search(this,_("Town"),"Town",0);
 }
 
 static void
+gui_internal_cmd_country(struct gui_priv *this, struct widget *wm, void *data)
+{
+	if (this->sl)
+		search_list_select(this->sl, attr_country_all, 0, 0);
+	gui_internal_search(this,_("Country"),"Country",1);
+}
+
+static void
 gui_internal_cmd_layout(struct gui_priv *this, struct widget *wm, void *data)
 {
 	struct attr attr;
@@ -2872,6 +2880,10 @@
 			image_new_l(this, "gui_town"), gravity_center|orientation_vertical,
 			gui_internal_cmd_town, NULL));
 	gui_internal_widget_append(w,
+		gui_internal_button_new_with_callback(this, _("Town (Intl.)"),
+			image_new_l(this, "gui_town"), gravity_center|orientation_vertical,
+			gui_internal_cmd_country, NULL));
+	gui_internal_widget_append(w,
 		gui_internal_button_new_with_callback(this, _("Quit"),
 			image_new_l(this, "gui_quit"), gravity_center|orientation_vertical,
 			gui_internal_cmd_quit, NULL));
