--- osm2navit.c.old	2008-06-07 09:42:31.000000000 +0200
+++ osm2navit.c	2008-06-07 09:43:50.000000000 +0200
@@ -149,7 +149,6 @@
 	"w	highway=bridleway	bridleway\n"
 	"w	highway=cycleway	cycleway\n"
 	"w	highway=cyclepath	cycleway\n"
-	"w	cycleway=track	cycleway\n"
 	"w	highway=footway		footway\n"
 	"w	highway=steps		steps\n"
 	"w	highway=unsurfaced	track_gravelled\n"
@@ -184,6 +183,7 @@
 	"w	highway=motorway	highway_city\n"
 	"w	highway=motorway_link	ramp\n"
 	"w	historic=town gate	poly_building\n"
+	"w	cycleway=track	        cycleway\n"
 	"w	amenity=parking		poly_car_parking\n"
 	"w	landuse=allotments	poly_allotments\n"
 	"w	landuse=cemetery	poly_cemetery\n"
