Index: navit/graphics/win32/graphics_win32.c
===================================================================
--- navit/graphics/win32/graphics_win32.c	(Revision 4053)
+++ navit/graphics/win32/graphics_win32.c	(Arbeitskopie)
@@ -70,6 +70,11 @@
 
 static HWND g_hwnd = NULL;
 
+#ifdef HAVE_API_WIN32_CE
+static int fullscr = 0;
+#endif
+
+
 #ifndef GET_WHEEL_DELTA_WPARAM
 #define GET_WHEEL_DELTA_WPARAM(wParam)  ((short)HIWORD(wParam))
 #endif
@@ -525,6 +530,15 @@
     case WM_KEYDOWN:
         HandleKeyDown( gra_priv, wParam);
         break;
+#ifdef HAVE_API_WIN32_CE
+    case WM_ENABLE:;
+        HWND hwndSip = FindWindow(L"MS_SIPBUTTON", NULL);
+        if (fullscr) {
+           // deactivate the SIP button
+           ShowWindow(hwndSip, SW_HIDE);
+        }
+        break;
+#endif
     default:
         return DefWindowProc(hwnd, Message, wParam, lParam);
     }
@@ -537,6 +551,7 @@
 #ifdef HAVE_API_WIN32_CE
     HWND hwndTaskbar = FindWindow(L"HHTaskBar", NULL);
     HWND hwndSip = FindWindow(L"MS_SIPBUTTON", NULL);
+    fullscr = on;
 	if (on) {
         ShowWindow(hwndTaskbar, SW_HIDE);
         MoveWindow(g_hwnd, 0, 0, GetSystemMetrics(SM_CXSCREEN), GetSystemMetrics(SM_CYSCREEN), FALSE);
Index: navit/osd/core/osd_core.c
===================================================================
--- navit/osd/core/osd_core.c	(Revision 4053)
+++ navit/osd/core/osd_core.c	(Arbeitskopie)
@@ -165,9 +165,16 @@
  * * @returns a pointer to a string containing the formatted speed
  * */
 static char * 
-format_speed(double speed, char *sep)
+format_speed(double speed, char *sep, char *format)
 {
-	return g_strdup_printf("%.0f%skm/h", speed, sep);
+	if (!format || !strcmp(format,"named"))
+		return g_strdup_printf("%.0f%skm/h", speed, sep);
+	else if (!strcmp(format,"value") || !strcmp(format,"unit")) {
+		if (!strcmp(format,"value"))
+			return g_strdup_printf("%.0f", speed);
+		else 
+			return "km/h";
+	} 
 }
 
 /*static char *
@@ -351,7 +358,7 @@
   }
 
   dist_buffer = format_distance(this->sum_dist,"");
-  spd_buffer = format_speed(spd,"");
+  spd_buffer = format_speed(spd,"","");
   remainder = this->time_all;
   days  = remainder  / (24*60*60);
   remainder = remainder  % (24*60*60);
@@ -1765,20 +1772,7 @@
 
 	switch (attr->type) {
 	case attr_position_speed:
-		if (!format || !strcmp(format,"named"))
-			return format_speed(*attr->u.numd,"");
-		if (!strcmp(format,"value") || !strcmp(format,"unit")) {
-			char *ret,*tmp=format_speed(*attr->u.numd," ");
-			char *pos=strchr(tmp,' ');
-			if (! pos)
-				return tmp;
-			*pos++='\0';
-			if (!strcmp(format,"value"))
-				return tmp;
-			ret=g_strdup(pos);
-			g_free(tmp);
-			return ret;
-		}
+		return format_speed(*attr->u.numd,"",format);
 	case attr_position_height:
 	case attr_position_direction:
 		return format_float_0(*attr->u.numd);
@@ -2039,7 +2033,6 @@
 					int *flags=tracking_get_current_flags(tracking);
 					if (flags && (*flags & AF_SPEED_LIMIT) && tracking_get_attr(tracking, attr_maxspeed, &maxspeed_attr, NULL)) {
 						routespeed = maxspeed_attr.u.num;
-						value = format_speed(routespeed, "");
 					}
 
 					if (routespeed == -1) {
@@ -2049,9 +2042,10 @@
 							rprof=vehicleprofile_get_roadprofile(prof, item->type);
 						if (rprof) {
 							routespeed=rprof->speed;
-							value=format_speed(routespeed,"");
 						}
 					}
+
+					value = format_speed(routespeed,"", oti->format);
 				} else if (item) {
 					if (tracking_get_attr(tracking, oti->attr_typ, &attr, NULL))
 						value=osd_text_format_attr(&attr, oti->format);
